### DISCUSSION TITLE
Use a-j/templ as the templating engine

### DISCUSSION THREAD

**Discussion by sawirricardo** [goravel/goravel#355] [Ideas]
Hi, I think it's cool that we can use this template engine in goravel, What do you guys think?
https://templ.guide/

**Comment by hwbrzzl** [comment]
If the package can be used with gin or fiber, it can be used with Goravel, too. It's very easy, but I don't try it, you can check this issue: https://github.com/a-h/templ/issues/349, you may get some good ideas.

**Comment by yanickrochon** [comment]
I propose a controller like this:

```go

import (
	view "goravel/resources/views"
)

func Web() {
	facades.Route().Get("/", func(ctx http.Context) http.Response {
		return ctx.Response().View().Template(view.Welcome("John"))
	})
}
```

Here's a showcase of a simple working implementation:

[![Templ integration proposal for Goravel](https://img.youtube.com/vi/1f9Kp2_Oo5Q/maxresdefault.jpg)](https://www.youtube.com/watch?v=1f9Kp2_Oo5Q)

**Reply by hwbrzzl** [reply]
Yes, it looks a good way, needs a deep investigation. There are no official plans to provide support for the time being, but welcome discussion and PR.

**Comment by AndreiTelteu** [comment]
I have a simpler working example and tutorial here:
https://dev.to/andreitelteu/how-to-use-templ-with-goravel-d63
It does not replace the configured goravel view engine so you can implement templ gradually on existing projects.

**Reply by hwbrzzl** [reply]
Amazing, great work 👍

**Reply by yanickrochon** [reply]
What I proposed can surely be improved. The question I would raise, though is: should the choice of rendering engine be built-in or done through external framework extensions (e.g., `"github.com/goravel/templ"`)?

In other words, should it be:

```go
func Web() {
  facades.Route().Get("/", func(ctx http.Context) http.Response {
    return ctx.Response().View().Templ(view.Welcome(/* ...args */))
  }
}
```

or

```go
func Web() {
  facades.Route().Get("/", func(ctx http.Context) http.Response {
    return RenderTempl(ctx, view.Welcome(/* ...args */))
  }
}
```

I would argue that, given the simplicity of integration, the first should be preferable because it follows the current framework pattern, and it still allows for opting in/out.

**Reply by hwbrzzl** [reply]
Yes, `ctx.Response().View().Template(view.Welcome(/* ...args */))` is good.

**Reply by neilyoung** [reply]
@AndreiTelteu Thanks for sharing. I recommend to edit the installation of air, since it has changed the repo in between:

go install github.com/air-verse/air@latest

**Reply by AndreiTelteu** [reply]
@neilyoung You are right. Updated. Thanks.
